#Author: Allison Haaning

#Script that combines GWAS results generated by GAPIT (Genome Association and Prediction Integrated Tool) R package for all traits. Writes two files, one containing results for all SNPs and all traits and another containing only markers significantly associated with all traits (FDR p-value < 0.05).

setwd(".")

#Get a list of file names for the GWAS results files in a directory 
GWAS_Results_Files_List <- list.files(path=".", pattern="*.GWAS.Results.csv")

#Subset the file names in the file names list to get short IDs for each file. To be used for naming data frames.
GWAS_Results_Files_IDs <- sub("GAPIT..","",GWAS_Results_Files_List)
GWAS_Results_Files_IDs <- sub(".GWAS.Results.csv","",GWAS_Results_Files_IDs)

#Get a list of file names for the allelic estimates file in the working directory
Allelic_Effects_Files_List <- list.files(path=".", pattern="*.Allelic*")

#Get a list of file names for the BIC model selection files in the working directory
BIC_Mod_Sel_Files_List <- list.files(path=".",pattern="*.BIC.Model.Selection.Results.csv")

#Iterate through the folder, read in all files in the two files lists, merge them together, and assign them to a data frame. The result will be a data frame for each GWAS.Results.csv file with a column of allelic effect estimates.
for (i in 1:length(GWAS_Results_Files_List)) assign (GWAS_Results_Files_IDs[i],merge(read.csv(GWAS_Results_Files_List[i]),read.csv(Allelic_Effects_Files_List[i]),by="SNP"))

#Merge all data into one data frame
All_GWAS_Results <- do.call(rbind, lapply(GWAS_Results_Files_IDs, function(x){data.frame(id=x, eval(parse(text=x)))}))

#Make a list of data frames by reading in all files with results from BIC model selection.
BIC_list <- lapply(BIC_Mod_Sel_Files_List,read.csv)

#Get the Number of PCs for which BIC was largest - this is the number of PCs used to correct for population structure.
BIC_Max <- lapply(BIC_list, function(x){x[which.max(x$BIC..larger.is.better....Schwarz.1978),1]})

#Add column of trait ids to number of PCs and change column names
BIC_Max <- data.frame(cbind(GWAS_Results_Files_IDs,unlist(BIC_Max)))
colnames(BIC_Max) <- c("id","PCs_for_Pop_Struct")

#Add column with the number of PCs to data frame containing all results
All_GWAS_Results <- merge(All_GWAS_Results,BIC_Max,by="id")

#Remove columns with duplicate info from merging
All_GWAS_Results$Position.y <- NULL
All_GWAS_Results$Chromosome.y <- NULL

#Rename columns where .x was added on during merging and rename id column
names(All_GWAS_Results)[names(All_GWAS_Results)=="Position.x"] <- "Position"
names(All_GWAS_Results)[names(All_GWAS_Results)=="Chromosome.x"] <- "Chromosome"
names(All_GWAS_Results)[names(All_GWAS_Results)=="id"] <- "Trait"

#Add on columns for date, lines, genotyping platform (GP), and other info related to results
All_GWAS_Results$Year <- 2014
All_GWAS_Results$Lines <- "All"
All_GWAS_Results$GP <- "50K_GBS_Merged_TAGS_R2_0.95"
All_GWAS_Results$Exp <- "Compressed_Mod_Sel"

save(All_GWAS_Results, file="All_GWAS_Results_2014_All_Lines_50K_GBS_Mod_Selection.RData")

sig_SNPs <- All_GWAS_Results[which(All_GWAS_Results$FDR_Adjusted_P.values<=0.05),]

save(sig_SNPs,file="Sig_GWAS_Results_2014_All_Lines_50K_GBS_Mod_Selection.RData")

